### DISCUSSION TITLE
What is the proper way to update custom attributes?

### DISCUSSION THREAD

**Discussion by Taegost** [Sellbrite/Sellbrite-API#11] [Q&A] [answered]
According to the [API documentation](https://developer.sellbrite.com/endpoints/products-sku-post), in the body we would include `custom_attributes: Array of Strings`, but that doesn't appear to be the way the data is represented.
In my case, I'm trying to update an attribute called "DnD Size Category", which is represented in the back-end as "dnd_size_category".

When I do a GET on a product that has this particular piece of data, this is what I get back (The remainder of the response is truncated for brevity):
```
"custom_attributes":
{
	"dnd_size_category":"test"
},
```

This doesn't appear to be an array at all; however since this is the only custom attribute assigned to this specific product, I'm willing to assume it will only return an array if there's more than one attribute. To take it one step further, it looks like a key-value pair rather than String as indicated in the documentation.

Regardless of whether I use POST, PUT, or PATCH, (using the same credentials as when I succeeded on the GET) I receive an error when I try to send up the following in the body:
```
"custom_attributes":
{
	"dnd_size_category":"Medium"
}
```

The error I'm getting back is:
```
"500 - "{\"error\":{\"message\":\"There was an error with your request. (undefined method `permit!' for \\\"{\\\\n\\\\t\\\\\\\"dnd_size_category\\\\\\\":\\\\\\\"Medium\\\\\\\"\\\\n}\\\":String)\"}}""
```

**Comment by justinmcdaniel** [comment]
I remember something being weird with this endpoint. I think you're right that it wants a dictionary; however, the other oddity I recall was that it didn't accept underscores in the key names. The underscores needed to be a [space] character.... even though the returned result uses underscores in the key names. 

Try using spaces instead of underscores in your POST.

EDIT: Only use spaces for the custom attribute keys. All other keys should have underscores.

i.e.

```json
{
  "sku": "your_sku",
  "custom_attributes": {
	"dnd size category":"Medium"
  }
}
```

**Reply by Taegost** [reply]
Thank you for your help! I'm still getting the same error, here are the new things I tried:

- I tried changing it up as you suggested, even including SKU in the body
- Tried using the same capitalization as I did in the GUI. 
- Removed spaces entirely (Tried both all lowercase and mixed-case as well)
- Created a new attribute without spaces

**Comment by cwilkins-godaddy** [comment]
It looks like the documentation is incorrect. The custom_attributes should receive an object. Could you try sending an object and report results here?

**Reply by Taegost** [reply]
What is the schema for the object? I've tried submitting it with several different combinations of key-value pair (See my response to @justinmcdaniel here: https://github.com/Sellbrite/Sellbrite-API/discussions/11#discussioncomment-6030702) with no success

**Reply by cwilkins-godaddy** [reply]
Looking back on the error in your initial post, I think you may be sending an escaped string for the value.

**Reply by Taegost** [reply]
The examples I gave in the original post came straight out of my API calls. I haven't tried anything fancier than a single alphanumeric word so far. I even tried the suggestions from @justinmcdaniel and didn't have any luck (see my response to him here: https://github.com/Sellbrite/Sellbrite-API/discussions/11#discussioncomment-5984205).
If I knew what the endpoint was actually expecting as a payload, I could do further testing, but at this point, I have no idea what the API is actually expecting me to send in this call.

**Reply by cwilkins-godaddy** [reply]
The value of custom_attributes should be an object.

**Comment by Taegost** [comment]
> The value of custom_attributes should be an object.
What is the schema for the object? I've tried submitting it with several different combinations of key-value pair (See my response to @justinmcdaniel here: https://github.com/Sellbrite/Sellbrite-API/discussions/11#discussioncomment-6030702) with no success

**Reply by cwilkins-godaddy** [reply]
@justinmcdaniel is correct about the spaces instead of underscores. That should be called out in the documentation.

The value of `custom_attributes` should be a json object:

```
"custom_attributes":
{
	"dnd size category":"Medium"
}
```

**Reply by Taegost** [reply]
As I stated previously, I tried that and still received the same error. I even tried removing all spaces entirely from the name of the attribute.

**Reply by cwilkins-godaddy** [reply]
Is this the error?:

```
"500 - "{\"error\":{\"message\":\"There was an error with your request. (undefined method `permit!' for \\\"{\\\\n\\\\t\\\\\\\"dnd_size_category\\\\\\\":\\\\\\\"Medium\\\\\\\"\\\\n}\\\":String)\"}}""
```

**Reply by Taegost** [reply]
Yes, that's the error I get. The only part that changes is what comes after `undefined method 'permit!' for` depending on what my test data is.
My latest test:
`"custom_attributes":
{
	"dndsizecategory":"Medium"
}`

Which gives me the response:

{"status":"rejected","reason":{"message":"500 - \"{\\\"error\\\":{\\\"message\\\":\\\"There was an error with your request. (undefined method `permit!' for \\\\\\\"{\\\\\\\\n\\\\\\\\t\\\\\\\\\\\\\\\"dndsizecategory\\\\\\\\\\\\\\\":\\\\\\\\\\\\\\\"Medium\\\\\\\\\\\\\\\"\\\\\\\\n}\\\\\\\":String)\\\"}}\"","name":"Error","stack":"Error: 500 - \"{\\\"error\\\":{\\\"message\\\":\\\"There was an error with your request. (undefined method `permit!' for \\\\\\\"{\\\\\\\\n\\\\\\\\t\\\\\\\\\\\\\\\"dndsizecategory\\\\\\\\\\\\\\\":\\\\\\\\\\\\\\\"Medium\\\\\\\\\\\\\\\"\\\\\\\\n}\\\\\\\":String)\\\"}}\"\n    at createError (/usr/local/lib/node_modules/n8n/node_modules/axios/lib/core/createError.js:16:15)\n    at settle (/usr/local/lib/node_modules/n8n/node_modules/axios/lib/core/settle.js:17:12)\n    at RedirectableRequest.handleResponse (/usr/local/lib/node_modules/n8n/node_modules/axios/lib/adapters/http.js:238:9)\n    at RedirectableRequest.emit (node:events:525:35)\n    at RedirectableRequest.emit (node:domain:489:12)\n    at RedirectableRequest._processResponse (/usr/local/lib/node_modules/n8n/node_modules/follow-redirects/index.js:356:10)\n    at ClientRequest.RedirectableRequest._onNativeResponse (/usr/local/lib/node_modules/n8n/node_modules/follow-redirects/index.js:62:10)\n    at Object.onceWrapper (node:events:628:26)\n    at ClientRequest.emit (node:events:525:35)\n    at ClientRequest.emit (node:domain:489:12)\n    at HTTPParser.parserOnIncomingClient (node:_http_client:693:27)\n    at HTTPParser.parserOnHeadersComplete (node:_http_common:128:17)\n    at TLSSocket.socketOnData (node:_http_client:534:22)\n    at TLSSocket.emit (node:events:513:28)\n    at TLSSocket.emit (node:domain:489:12)\n    at addChunk (node:internal/streams/readable:315:12)\n    at readableAddChunk (node:internal/streams/readable:289:9)\n    at TLSSocket.Readable.push (node:internal/streams/readable:228:10)\n    at TLSWrap.onStreamRead (node:internal/stream_base_commons:190:23)"}}

**Reply by cwilkins-godaddy** [reply]
Can you show the complete log of the http request? It looks like the object value of custom attributes is being serialized to string, either on your end or my end.

**Reply by Taegost** [reply]
I use n8n for my automation, here's the relevant output it gives me for the request being sent:

```
method: 'PUT', uri: 'https://api.sellbrite.com/v1/products/DM-32-00-aboro-ko-cleric-tears', gzip: true, rejectUnauthorized: true, …}
body: 
custom_attributes: "{\"dndsizecategory\":\"Medium\"}"
```

**Reply by cwilkins-godaddy** [reply]
Yeah, it's sending the value as a string.

**Reply by Taegost** [reply, marked as answer]
I was able to get it working! When I saw that it was escaping the quotes I changed the way I was passing the parameter (I gave the exact JSON rather than let n8n do it for me) and it worked. It totally trampled the attributes that were already there and replaced them with only the single attribute I passed in, but it worked. Luckily, it was just test data.
Now, how do we go about getting the documentation updated so that it shows `custom_attributes` take key-value-pairs rather than just a string?
